armory/baseline_models/keras/keras_resnet50.py

```diff
@@ -1,7 +1,29 @@
+import numpy as np
 from art.classifiers import KerasClassifier
-from tensorflow.keras.applications.resnet50 import ResNet50, preprocess_input
+from tensorflow.keras.applications.resnet50 import ResNet50
+from tensorflow.keras.preprocessing import image
 
-preprocessing_fn = preprocess_input
+
+def preprocess_input_resnet50(img):
+    # Model was trained with inputs zero-centered on ImageNet mean
+    mean = [103.939, 116.779, 123.68]
+    img = img[..., ::-1]
+    img[..., 0] -= mean[0]
+    img[..., 1] -= mean[1]
+    img[..., 2] -= mean[2]
+
+    return img
+
+
+def preprocessing_fn(x: np.ndarray) -> np.ndarray:
+    shape = (224, 224)  # Expected input shape of model
+    output = []
+    for i in range(x.shape[0]):
+        im_raw = image.array_to_img(x[i])
+        im = image.img_to_array(im_raw.resize(shape))
+        output.append(im)
+    output = preprocess_input_resnet50(np.array(output))
+    return output
 
 
 def get_art_model(model_kwargs, wrapper_kwargs):
```

Comment:
What should we set the clip_values to here? There is a different upper-bound for each channel.

Reply:
It's really hard for me to say. 
https://github.com/keras-team/keras-applications/blob/master/keras_applications/imagenet_utils.py
^ has a lot of different processing paths.

It depends on what the original model was trained on. They are zero-centered on ImageNet mean. Are they also scaled? Ideally, we would take the largest inputs from the original domain for each channel ([0, 255], I think?), then apply the preprocessing to get the range for each channel. If the per-channel domains are different, then we'll need to pass in a tuple of np.arrays to the clip_values kwarg.

Reply:
So I took the preprocessing functions for each model and pulled them out of keras.applications to limit confusion. This particular model uses:

I didn't know clip_values could be set per dimension, [but it looks as though they can](https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/classifiers/classifier.py#L47-L50). PTAL at the changes.

Reply:
I'll need to play around with the changes to ensure this is what ART expects, but this looks good.

Reply:
Yes, this looks good. 

ART uses `np.clip` under the hood, so it broadcasts the clip values.